varnish/plan.sh

```diff
@@ -9,10 +9,31 @@ pkg_version="5.1.2"
 pkg_source="https://varnish-cache.org/_downloads/${pkg_name}-${pkg_version}.tgz"
 
 pkg_shasum="39d858137e26948a7c85f07363f13f0778da61d234126e03a160a0cb9ba4fce3"
-pkg_deps=(core/glibc core/ncurses core/docutils core/pcre core/gcc core/bash )
-pkg_build_deps=(core/make core/python2 core/pkg-config core/readline core/graphviz core/libtool core/libedit core/automake core/m4 core/autoconf)
+pkg_deps=(
```

Comment:
Huge 👍 for single lining these, makes diffs super easy to read.   

However, I don't think alphabetizing them is the right thing to do.  Order for `pkg_deps` and `pkg_build_deps` is important.  While it's unlikely to cause an issue in this case, by alphabetizing them we're implying that they should always be alphabetized which is not the case.

Reply:
why does order matter? They all get installed together in a single phase of the build lifecycle, and the transitive deps don't matter.

Reply:
Order determines how the various PATH variables are constructed.  I've seen with other plans that alphabetizing can cause introduce a change in behavior.  Usually busybox is the culprit as it comes before most every other plan.  No busybox here which is why I _think_ we're safe, but I can't recall if there are any other cases.  

This is probably a documentation/discovery problem around how PATH et. al are constructed AND worst case we can revert and build so I'm not going to 👎 the alphabetization, but my preference would be to leave order as-is.

Reply:
I'm with @smacfarlane on this. I've seen pathing issues with order changes. Alphabetization is something I've done on a couple of plans that I no longer do, because dragons.

Reply:
I think specifically "don't do that because it breaks and we don't know why" is _exactly_ the reason that we *SHOULD* be breaking things.

Essentially, I think that order shouldn't matter otherwise we're relying on "magic" and possibly poorly documented reasons for ordering.  Which creates a poor developer experience.  Also, I'm generally opposed to cargo-cult programming.

Heck, how many times have you encountered the "we removed this comment and now our code won't compile"?

If ordering does matter, then I think there should be some method/function for explicitly calling this out rather than saying "the order they're included in the include list is the order".  

E.g.: if busybox needs to be included first, we should 1) understand why, and 2) ensure that if busybox is included in the dep list, it comes first in the include line, regardless of the order of packages.

I'm totally willing to have my mind changed 😸 and I'm not saying I'm right... but from my perspective, making the developer experience as amazing as possible is what's really going to help drive adoption.

Reply:
I hadn't even thought of that! Good point. Thanks for explaining!

Reply:
@qubitrenegade Thanks for having this discussion,  I think we're getting some really useful feedback out of it and I 1000% agree on the developer experience.  

While I agree with you on cargo-culting and "magic", I don't think that's the case in this instance.  

We know how Habitat (specifically hab-plan-build) constructs the paths, and it is in a deterministic manner.  The complexity and unknowns come from the combination of packages in the pkg_deps and pkg_build_deps, what bins/libs/headers they provide, the build process of the target package itself and how it interacts with other shell commands at run time.   

Made up example: Consider package `foo` with `pkg_build_deps=(a/bar a/baz)`.  Both `a/bar` and `a/baz` have a `pkg_include_dirs=(include)` and both have a headers: `include/sys/helpers.h`.  The order in which you include your build deps determines which header you will get if `foo` has a file with `#include "sys/helpers.h"`. This could cause build failures due to function signature mis-matches or unimplemented functions.  In the worst case, both bar and baz implement the same functions, with the same signatures, but with different behavior.  

Specifically with busybox,  it implements a lot of standard unix tools that don't support all the features you may be used to, or they have different default behaviors. If you include it alphabetically, it will almost always be first and you may not get the commands you think you're getting so it should almost always be last. 

This is why I think it's not a  "magic" or cargo cult issue, but one of complexity due to the number of packages, what those packages provides and ways they can be combined.